Fix for distributed tracing instrumentation on blocking controllers

lgathy opened the pull request:
Closes #4383 

The problem seemed to be that the resultEmitter is constructed before the http filter chain was assembled and executed, thus the deffered Flowable could not catch all the changes to the context done by any of the filters or the publishers created by them. I think the only valid option is to delay this defer until the subscribe() of the publisher produced by the innermost http filter is called. This is only done after subscribing to all publishers created by all the other http filters. However, I needed to add an extra `Flowable.defer(...)` operation to achive this. I am not sure if this is the best way, but it does not seem to break any other tests.

lgathy committed:
Demonstrate Jaeger tracing bug in blocking controller methods

lgathy committed:
Deferring the execution of the controller method happened before the TracingPublisher.onSubscribe(...)

lgathy commented:
@ilopmar @graemerocher Is it possible to merge this before the `2.3.2` release, if you agree with the approach I have taken? Luckily, only one private method signature change was needed.

jameskleeh commented:
A broken test showing the issue would be a better place to start. At a glance these changes don't make sense to me

lgathy commented:
@jameskleeh The first commit changes the built-in test so that it breaks. Isn't that sufficient enough?

jameskleeh commented:
@lgathy If the existing tests break then I would assume that existing users applications might break. Do you think that is the case? If you need to change the tests in order for them to break then you are effectively no longer testing the previous case and thus new tests should be added instead of changing old ones.

graemerocher commented:
@lgathy Thanks for the PR!

 I think the change makes sense however I agree with @jameskleeh that we need to verify the other case works as well so we would need an additional set of tests. Would you be able to modify the test to test both cases? Thanks.

lgathy commented:
@jameskleeh Ok, I get your point.

Actually I see these changes as fixing that test to actually test what its original intention was. Those controller methods intended to be blocking methods back in `1.x`, however in `2.0.0` [this breaking change](https://docs.micronaut.io/latest/guide/index.html#_thread_pool_selection_now_manual) was introduced, and since this test controller was not adopted to that change those methods were no longer executed on the IO threads as before. So applying the `@ExecuteOn` annotations only makes them `2.x` compatible.

Nonetheless, we could create new additional test cases if you say that's better. Honestly I thought that having a blocking method without the `@ExecuteOn` annotation is actually misusage so that's why I went for adding the annotation on the existing methods instead. What do you think?

lgathy commented:
@graemerocher Sure, I will do that. Thanks for the feedback.

graemerocher commented:
@lgathy I understand your point though in just because a method returns `String` doesn't mean it blocks the event loop. Some of the methods in the test actually do block the event loop but others do not I/O so don't block. I think having a set of tests that cover both cases would be a good thing.

lgathy committed:
Adding separate tracing test cases for blocking an non-blocking controller methods

lgathy commented:
@graemerocher That is indeed a valid use case, I slipped through that and I also didn't notice that @jameskleeh was refering to that still being a valid use-case, tbh.

I modified the PR now, reverted the changes to the existing methods and added separate ones for the blocking use-case alongside with test cases. There would be a couple more possible combinations - i.e. to call blocking from non-blocking and vica versa in the nested test cases. I am not sure if those would make any difference in the execution. Do you think we would need them?

graemerocher commented:
I think this is probably enough for now. Thanks for the contribution
